bindings/python/src/utils/pretokenization.rs

```diff
@@ -0,0 +1,256 @@
+use tokenizers as tk;
+
+use pyo3::exceptions;
+use pyo3::prelude::*;
+use pyo3::types::*;
+
+use super::{
+    DestroyPtr, PyNormalizedString, PyNormalizedStringRefMut, RefMutContainer, RefMutGuard,
+};
+use crate::encoding::PyEncoding;
+use crate::error::ToPyResult;
+use crate::token::PyToken;
+use tk::{OffsetReferential, OffsetType, Offsets, PreTokenizedString, Token};
+
+fn split(pretok: &mut PreTokenizedString, func: &PyAny) -> PyResult<()> {
+    if !func.is_callable() {
+        Err(exceptions::PyTypeError::new_err(
+            "`split` expect a callable with the signature: \
+           `fn(normalized: NormalizedString) -> List[NormalizedString]`",
+        ))
+    } else {
+        ToPyResult(pretok.split(|i, normalized| {
+            let output = func.call((i, PyNormalizedString::from(normalized)), None)?;
+            Ok(output
+                .extract::<Vec<PyNormalizedString>>()?
+                .into_iter()
+                .map(tk::NormalizedString::from))
+        }))
+        .into()
+    }
+}
+
+fn normalize(pretok: &mut PreTokenizedString, func: &PyAny) -> PyResult<()> {
+    if !func.is_callable() {
+        Err(exceptions::PyTypeError::new_err(
+            "`normalize` expect a callable with the signature: \
+            `fn(normalized: NormalizedString)`",
+        ))
+    } else {
+        ToPyResult(pretok.normalize(|normalized| {
+            let norm = PyNormalizedStringRefMut::new(normalized);
+            func.call((norm.get(),), None)?;
+            Ok(())
+        }))
+        .into()
+    }
+}
+
+fn tokenize(pretok: &mut PreTokenizedString, func: &PyAny) -> PyResult<()> {
+    if !func.is_callable() {
+        Err(exceptions::PyTypeError::new_err(
+            "`tokenize` expect a callable with the signature: \
+            `fn(str) -> List[Token]`",
```

Comment:
You had the index of the token in the example, no ?

Reply:
Like with the `normalizer` the api surface seems a bit too much. `pretok.split(Fn[PretokenizedString] -> List[PretokenizedString])` is general enough, no ?

Reply:
Indeed, split takes the index also, nice catch!